### PR TITLE
Parse box pattern syntax

### PR DESCRIPTION
**Pull request by MSleepyPanda**
Fixes https://github.com/shepmaster/fuzzy-pickles/issues/93

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -2855,6 +2865,7 @@ shims! [
     (asterisk, Token::into_asterisk, Error::ExpectedAsterisk),
     (at, Token::into_at, Error::ExpectedAt),
     (bang, Token::into_bang, Error::ExpectedBang),
+    (box_pattern, Token::into_box, Error::ExpectedBox),
```
Comment (shepmaster) [inline src/lib.rs]: You can just reuse `kw_box`, defined above (`kw_` is to avoid conflict with the *real* keyword) :-)

File: src/lib.rs [hunk 2]
```diff
@@ -5337,6 +5361,12 @@ mod test {
     }
 
     #[test]
+    fn fn_with_arguments_with_box_pattern() {
```
Comment (shepmaster) [inline src/lib.rs]: I think it's safe to remove this test; we already cover that functions use the general set of "pattern" above. The presence of this test makes it look like there's something extra-special about box patterns in function arguments, which there isn't.

File: src/lib.rs [hunk 3]
```diff
@@ -1631,6 +1631,7 @@ pub enum PatternKind {
     String(PatternString),
     Struct(PatternStruct),
     Tuple(PatternTuple),
+    Box(PatternBox)
```
Comment (shepmaster) [inline src/lib.rs]: Go ahead and place this alphabetically in the list, please.

File: src/lib.rs (hunk 1 above)
Comment (MSleepyPanda) [reply src/lib.rs]: Ah dang, i knew it would be there but must have missed it! :D

**Comment by MSleepyPanda** [comment]
Addressed your comments, should be fine now :) Thanks!

**Comment by shepmaster** [comment]
Superb! Thanks again!

If you don't mind me asking, how have you found using this library? Are you using it for anything exciting you want to share?

**Comment by MSleepyPanda** [comment]
Exams are over and i finally have time to rewrite [explain-rs](https://www.reddit.com/r/rust/comments/6l6klk/community_automation_explaining_rust_syntax/). (Rewrite because syntex is no longer maintained). 

Why i didn't choose syn / choose yours:
 * Span support is somewhat unstable and tied to the compiler
 * Error messages, somewhat similar to the compiler
 * Since your use case is tailored for analysis, i hope to benefit from overall more ergnomic things. I've experienced a few papercuts when using syn, but that makes sense, considering that syns main usecase is deriving

Your crate is mentioned though strata-rust several times in the issues of syn, mostly as a ["this is how it should be done"](https://github.com/dtolnay/syn/issues/136) example.

**Comment by shepmaster** [comment]
Cool! As far as I know, I'm the only other user of this crate, so I'm probably blind to my own set of papercuts. Feel free to file an issue if you find any.

**Pull request merged**